File: solana_launchpads/launchpad_decoder/adapters/pumpfun.py
"""pump.fun bonding curves.

Program: 6EF8rrecthR5Dkzon8Nwu78hRvfCKubJ14M5uBEwF6P

Every launch parameter is program state, not a constant: the single `Global`
account (PDA of ["global"]) holds `initial_virtual_token_reserves`,
`initial_virtual_sol_reserves`, `initial_real_token_reserves` and
`token_total_supply`.  Read those and the launch price, launch market cap and
the 85-SOL-ish raise target all fall out of `x*y=k`.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from .. import curves
from ..types import CurveFamily, LaunchMetrics, ValueSource, price_raw_to_ui, ui_amount
from .base import Adapter, DecodeContext

GLOBAL_ADDRESS = "4wTV1YmiEkRvAtNtsSGPtUrqRYQMe5SKy2uB4Jjaxnjf"

#: Values read off mainnet `Global` and published in pump-fun/pump-public-docs.
#: Used only when no account source is available; anything sourced from here is
#: tagged BUNDLED_SNAPSHOT so a consumer can tell it was not read from chain.
GLOBAL_SNAPSHOT: Dict[str, int] = {
    "initial_virtual_token_reserves": 1_073_000_000_000_000,
    "initial_virtual_sol_reserves": 30_000_000_000,
    "initial_real_token_reserves": 793_100_000_000_000,
    "token_total_supply": 1_000_000_000_000_000,
    "fee_basis_points": 100,
}


def pick(state: Dict[str, Any], *names: str) -> Optional[Any]:
    """First present, non-None field -- absorbs field renames across versions."""
    for name in names:
        if state.get(name) is not None:
            return state[name]
    return None


class PumpFunAdapter(Adapter):
    key = "pumpfun"
    curve_family = CurveFamily.CONSTANT_PRODUCT_VIRTUAL

    def global_config(self, ctx: DecodeContext):
        name, decoded, slot = ctx.configs.fetch(
            ctx.spec.program_id, GLOBAL_ADDRESS, ctx.schema
        )
        if decoded is not None:
            return decoded, ValueSource.ONCHAIN_CONFIG, slot
        return dict(GLOBAL_SNAPSHOT), ValueSource.BUNDLED_SNAPSHOT, 0

    def decode(
        self,
        ctx: DecodeContext,
        account_name: str,
        state: Dict[str, Any],
        address: Optional[str] = None,
    ) -> LaunchMetrics:
        m = self.new_metrics(ctx, account_name, address, state)
        cfg, cfg_source, cfg_slot = self.global_config(ctx)

        quote_mint = state.get("quote_mint")
        if not quote_mint or quote_mint == "11111111111111111111111111111111":
            quote_mint = ctx.spec.default_quote_mint
        m.quote_mint = quote_mint
        m.creator = state.get("creator")

        m.base_decimals = ctx.spec.default_base_decimals or 6
        m.quote_decimals = ctx.decimals_for(quote_mint) or 9

        v_base = pick(state, "virtual_token_reserves")
        v_quote = pick(state, "virtual_quote_reserves", "virtual_sol_reserves")
        real_base = pick(state, "real_token_reserves")
        real_quote = pick(state, "real_quote_reserves", "real_sol_reserves")
        total_supply_raw = pick(state, "token_total_supply") or cfg.get("token_total_supply")

        # -- supply -------------------------------------------------------
        m.total_supply = self.param(
            ui_amount(total_supply_raw, m.base_decimals),
            ValueSource.ONCHAIN_STATE if state.get("token_total_supply") else cfg_source,
            ctx.slot,
        )

        init_v_base = cfg.get("initial_virtual_token_reserves")
        init_v_quote = pick(
            cfg, "initial_virtual_sol_reserves", "initial_virtual_quote_reserves"
        )
        init_real_base = cfg.get("initial_real_token_reserves")

        m.tokens_for_sale = self.param(
            ui_amount(init_real_base, m.base_decimals), cfg_source, cfg_slot
        )
        if real_base is not None and init_real_base is not None:
            m.tokens_sold = self.param(
                ui_amount(max(0, init_real_base - real_base), m.base_decimals),
                ValueSource.DERIVED,
                ctx.slot,
                "initial_real_token_reserves - real_token_reserves",
            )

        # -- prices -------------------------------------------------------
        if init_v_base and init_v_quote:
            m.launch_price_quote = self.param(
                price_raw_to_ui(
                    curves.cp_price_raw(init_v_quote, init_v_base),
                    m.base_decimals,
                    m.quote_decimals,
                ),
                cfg_source,
                cfg_slot,
                "initial_virtual_quote / initial_virtual_token",
            )
        if v_base and v_quote is not None:
            m.current_price_quote = self.param(
                price_raw_to_ui(
                    curves.cp_price_raw(v_quote, v_base), m.base_decimals, m.quote_decimals
                ),
                ValueSource.ONCHAIN_STATE,
                ctx.slot,
                "virtual_quote_reserves / virtual_token_reserves",
            )

        # -- raise target and graduation price ----------------------------
        if init_v_base and init_v_quote and init_real_base:
            target_raw = curves.cp_raise_for_supply(init_v_quote, init_v_base, init_real_base)
            m.raise_target_quote = self.param(
                ui_amount(target_raw, m.quote_decimals),
                ValueSource.DERIVED,
                cfg_slot,
                "quote needed to drain initial_real_token_reserves (pre-fee)",
            )
            final_price_raw = curves.cp_final_price_raw(
                init_v_quote, init_v_base, init_real_base
            )
            m.graduation_price_quote = self.param(
                price_raw_to_ui(final_price_raw, m.base_decimals, m.quote_decimals),
                ValueSource.DERIVED,
                cfg_slot,
                "spot price once real_token_reserves hits zero",
            )

        if real_quote is not None:
            m.raised_quote = self.param(
                ui_amount(real_quote, m.quote_decimals), ValueSource.ONCHAIN_STATE, ctx.slot
            )

        complete = state.get("complete")
        m.complete = self.param(complete, ValueSource.ONCHAIN_STATE, ctx.slot)
        if complete:
            m.migrated = self.param(
                True, ValueSource.DERIVED, ctx.slot, "curve complete -> migrates to PumpSwap"
            )

        fee_bps = cfg.get("fee_basis_points")
        creator_fee = pick(state, "creator_fee_bps") or cfg.get("creator_fee_basis_points")
        if fee_bps is not None:
            m.fee_bps = self.param(
                fee_bps + (creator_fee or 0),
                cfg_source,
                cfg_slot,
                "protocol fee_basis_points + creator fee",
            )

        if state.get("_truncated_fields"):
            m.warn(
                "account predates the current program layout; missing fields: "
                + ", ".join(state["_truncated_fields"])
            )
        if cfg_source is ValueSource.BUNDLED_SNAPSHOT:
            m.warn(
                "Global config was not read from chain -- launch parameters come "
                "from a bundled snapshot and may be stale"
            )
        return self.finish(m)

File: solana_launchpads/go/launchpad/pumpfun.go
package launchpad

// pump.fun -- program 6EF8rrecthR5Dkzon8Nwu78hRvfCKubJ14M5uBEwF6P
//
// Every launch parameter is program state, not a constant: the single Global
// account (PDA of ["global"], 4wTV1YmiEkRvAtNtsSGPtUrqRYQMe5SKy2uB4Jjaxnjf)
// holds the initial virtual reserves, the sellable supply and the total supply.
// Read those and the launch price, market cap and raise target all fall out of
// x*y=k.
//
// The four leading u64s sit at account offsets 8/16/24/32 in every generation
// of this account, verified against all 16 published revisions of the IDL. The
// only change at those offsets was a rename in the "USDC paired coins"
// revision: virtual_sol_reserves -> virtual_quote_reserves and
// real_sol_reserves -> real_quote_reserves. Same bytes, same positions -- but
// after it "quote" appears in both the virtual and the real field names, which
// is an easy way to wire the two together wrongly.

const (
	PumpfunProgramID = "6EF8rrecthR5Dkzon8Nwu78hRvfCKubJ14M5uBEwF6P"
	PumpfunGlobalPDA = "4wTV1YmiEkRvAtNtsSGPtUrqRYQMe5SKy2uB4Jjaxnjf"

	// pump.fun mints at 6 decimals.
	PumpfunBaseDecimals = 6
)

// PumpfunParams holds everything derived from the Global account. Build it once
// per config and reuse it for every curve: the raise target and graduation
// price need 128-bit division, and neither depends on the curve.
type PumpfunParams struct {
	InitialVirtualBase  uint64
	InitialVirtualQuote uint64
	InitialRealBase     uint64
	TokenTotalSupply    uint64
	FeeBasisPoints      uint64

	// Variants holds every opening quote reserve the program seeds curves
	// with. pump.fun uses one for SOL-quoted coins and another for non-SOL
	// ones, and the curve account does not say which it is -- the difference
	// between its virtual and real quote reserve does.
	Variants VariantCandidates

	// derived once
	RaiseTargetRaw     uint64
	LaunchPriceRaw     float64
	GraduationPriceRaw float64
	QuoteDecimals      int
	// FromChain is false when the params were not read from a Global account.
	FromChain bool
}

// NewPumpfunParams derives the reusable constants from a decoded Global.
//
// quoteDecimals is 9 for SOL-quoted coins. pump.fun also supports non-SOL
// quotes, whose opening reserve is Global.InitialVirtualQuoteReserves rather
// than InitialVirtualSolReserves and whose quote mint is typically 6 decimals;
// mixing the two up scales every price by a thousand.
func NewPumpfunParams(g *PumpfunGlobal, quoteDecimals int) PumpfunParams {
	p := PumpfunParams{
		InitialVirtualBase:  g.InitialVirtualTokenReserves,
		InitialVirtualQuote: g.InitialVirtualSolReserves,
		InitialRealBase:     g.InitialRealTokenReserves,
		TokenTotalSupply:    g.TokenTotalSupply,
		FeeBasisPoints:      g.FeeBasisPoints,
		QuoteDecimals:       quoteDecimals,
		FromChain:           true,
		Variants: VariantCandidates{
			"sol":   g.InitialVirtualSolReserves,
			"quote": g.InitialVirtualQuoteReserves,
		},
	}
	p.derive()
	return p
}

// curveTypeName avoids concatenating on the hot path: the two real variants
// and the two fallbacks are constants, so pricing a curve allocates nothing.
func curveTypeName(variant string) string {
	switch variant {
	case "sol":
		return "constant_product:sol"
	case "quote":
		return "constant_product:quote"
	case "":
		return "constant_product:unclassified"
	}
	return "constant_product:" + variant
}

func (p *PumpfunParams) derive() {
	if p.InitialVirtualBase == 0 || p.InitialVirtualQuote == 0 {
		return
	}
	p.LaunchPriceRaw = CPPriceRaw(p.InitialVirtualQuote, p.InitialVirtualBase)
	if target, ok := CPRaiseForSupply(p.InitialVirtualQuote, p.InitialVirtualBase, p.InitialRealBase); ok {
		p.RaiseTargetRaw = target
	}
	if final, ok := CPFinalPriceRaw(p.InitialVirtualQuote, p.InitialVirtualBase, p.InitialRealBase); ok {
		p.GraduationPriceRaw = final
	}
}

// Metrics turns one decoded bonding curve into the unified answer.
//
// truncatedAt is what DecodeAccount reported; pass "" when the account decoded
// in full.
func (p PumpfunParams) Metrics(c *PumpfunBondingCurve, truncatedAt string) Metrics {
	baseDec, quoteDec := PumpfunBaseDecimals, p.QuoteDecimals
	if quoteDec == 0 {
		quoteDec = 9
	}

	// Which opening was this curve seeded with? Read it off the reserves
	// rather than assuming, or a non-SOL-quoted curve is priced against the
	// SOL opening and every number comes out wrong.
	variant, _ := ClassifyVariant(c.VirtualQuoteReserves, c.RealQuoteReserves, p.Variants)
	derived := p
	curveType := curveTypeName(variant)

	// When the difference matches no constant the program uses TODAY, the curve
	// still states its own opening -- the difference IS that opening. Prefer it
	// over the config's: a curve seeded under an older or per-mint constant is
	// correct, and judging it against today's Global prices it wrong and then
	// reports the correct row as broken.
	opening := RecoverOpening(
		c.VirtualTokenReserves, c.VirtualQuoteReserves,
		c.RealTokenReserves, c.RealQuoteReserves)
	measured := false
	if variant == "" {
		if opening.QuoteSeed > 0 {
			measured = true
			curveType = "constant_product:measured"
			quote := uint64(opening.QuoteSeed)
			base := opening.InitialVirtualBase()
			realBase := opening.ImpliedInitialRealBase
			// A consistent curve recovers exactly the opening the params
			// already hold, so re-deriving would be identical work for an
			// identical answer -- and re-deriving is the only thing on this
			// path that allocates.
			if quote != p.InitialVirtualQuote ||
				(base > 0 && base != p.InitialVirtualBase) ||
				(realBase > 0 && realBase != p.InitialRealBase) {
				derived.InitialVirtualQuote = quote
				if base > 0 {
					derived.InitialVirtualBase = base
				}
				if realBase > 0 {
					derived.InitialRealBase = realBase
				}
				derived.derive()
			}
		} else {
			curveType = "constant_product:unclassified"
		}
	} else if p.Variants[variant] != p.InitialVirtualQuote {
		derived.InitialVirtualQuote = p.Variants[variant]
		derived.derive()
	}

	paramsSource := SourceOnchainConfig
	if !p.FromChain {
		paramsSource = SourceFallback
	}

	totalSupply := c.TokenTotalSupply
	if totalSupply == 0 {
		totalSupply = p.TokenTotalSupply
	}

	m := Metrics{
		Launchpad:     "pumpfun",
		ProgramID:     PumpfunProgramID,
		Family:        FamilyConstantProductVirtual,
		CurveType:     curveType,
		BaseMint:      Pubkey{},
		QuoteMint:     c.QuoteMint,
		BaseDecimals:  baseDec,
		QuoteDecimals: quoteDec,
		TotalSupply:   UIAmount(totalSupply, baseDec),
		TokensForSale: UIAmount(derived.InitialRealBase, baseDec),
		Raised:        UIAmount(c.RealQuoteReserves, quoteDec),
		RaiseTarget:   UIAmount(derived.RaiseTargetRaw, quoteDec),
		Complete:      c.Complete,
		Migrated:      c.Complete,
		FeeBps:        float64(p.FeeBasisPoints + c.CreatorFeeBps),
		PriceSource:   SourceOnchainState,
		ParamsSource:  paramsSource,
		TruncatedAt:   truncatedAt,
	}
	m.LaunchPrice = PriceUI(derived.LaunchPriceRaw, baseDec, quoteDec)
	m.GraduationPrice = PriceUI(derived.GraduationPriceRaw, baseDec, quoteDec)
	if c.VirtualTokenReserves > 0 {
		m.CurrentPrice = PriceUI(
			CPPriceRaw(c.VirtualQuoteReserves, c.VirtualTokenReserves), baseDec, quoteDec)
	}
	if derived.InitialRealBase >= c.RealTokenReserves {
		m.TokensSold = UIAmount(derived.InitialRealBase-c.RealTokenReserves, baseDec)
	}

	// Does the pair actually lie on the curve? A pair that is individually
	// plausible but whose product is not k means the two reserves did not come
	// from the same read, and nothing else would catch it.
	//
	// Two kinds of wrong, and only one of them is the row's fault. A pair that
	// is not on ITS OWN curve is broken. A pair that is on its own curve but
	// not on today's default opening is merely unfamiliar, and saying otherwise
	// rejects every curve the program seeded differently.
	m.Violations = CheckCurveOpening(opening, totalSupply)
	if measured {
		// The opening came off the curve itself, so comparing k against it is
		// circular -- it holds by construction. All that is left to say is
		// which opening this curve claims, so it can be bucketed.
		m.Violations = append(m.Violations,
			CheckReserveOffset(c.VirtualQuoteReserves, c.RealQuoteReserves, p.Variants)...)
	} else {
		m.Violations = append(m.Violations, CheckConstantProductPair(
			c.VirtualTokenReserves, c.VirtualQuoteReserves,
			derived.InitialVirtualBase, derived.InitialVirtualQuote)...)
	}
	if !m.OK() {
		m.PriceSource = "SUSPECT"
		m.SuspectField, _ = DiagnosePair(
			c.VirtualTokenReserves, c.VirtualQuoteReserves,
			derived.InitialVirtualBase, derived.InitialVirtualQuote, derived.InitialRealBase)
	}

	m.fillDerived()
	return m
}
